File: backend/batch/crawl/contest.go
package crawl

import (
	"context"
	"fjnkt98/atcodersearch/pkg/atcoder"
	"fmt"
	"slices"

	"github.com/doug-martin/goqu/v9"
	_ "github.com/doug-martin/goqu/v9/dialect/postgres"
	"github.com/jackc/pgx/v5/pgxpool"
)

func SaveContests(ctx context.Context, pool *pgxpool.Pool, contests []atcoder.Contest) (int64, error) {
	if len(contests) == 0 {
		return 0, nil
	}

	tx, err := pool.Begin(ctx)
	if err != nil {
		return 0, fmt.Errorf("begin transaction: %w", err)
	}
	defer tx.Rollback(ctx)

	dialect := goqu.Dialect("postgres")

	var count int64 = 0
	for chunk := range slices.Chunk(contests, 1000) {
		q := dialect.Insert("contests").Prepared(true).Cols(
			"contest_id",
			"start_epoch_second",
			"duration_second",
			"title",
			"rate_change",
			"category",
			"updated_at",
		)
		for _, c := range chunk {
			q = q.Vals(goqu.Vals{
				c.ID,
				c.StartEpochSecond,
				c.DurationSecond,
				c.Title,
				c.RateChange,
				c.Categorize(),
				goqu.L("NOW()"),
			})
		}

		q = q.OnConflict(
			goqu.DoUpdate(
				"contest_id",
				goqu.Record{
					"contest_id":         goqu.L("EXCLUDED.contest_id"),
					"start_epoch_second": goqu.L("EXCLUDED.start_epoch_second"),
					"duration_second":    goqu.L("EXCLUDED.duration_second"),
					"title":              goqu.L("EXCLUDED.title"),
					"rate_change":        goqu.L("EXCLUDED.rate_change"),
					"category":           goqu.L("EXCLUDED.category"),
					"updated_at":         goqu.L("NOW()"),
				},
			),
		)

		sql, args, err := q.ToSQL()
		if err != nil {
			return 0, fmt.Errorf("to sql: %w", err)
		}

		result, err := tx.Exec(ctx, sql, args...)
		if err != nil {
			return 0, fmt.Errorf("exec: %w", err)
		}

		count += result.RowsAffected()
	}

	if err := tx.Commit(ctx); err != nil {
		return 0, fmt.Errorf("commit transaction: %w", err)
	}

	return count, nil
}

File: backend/src/crawl/contest.rs
use crate::atcoder::{AtCoderProblemsClient, Contest};
use anyhow::Context;
use sqlx::{self, postgres::Postgres, Acquire, Pool, QueryBuilder};

pub async fn crawl_contests(
    client: &AtCoderProblemsClient,
    pool: &Pool<Postgres>,
) -> anyhow::Result<()> {
    let contests = client
        .fetch_contests()
        .await
        .with_context(|| "crawl contest")?;

    let mut tx = pool.begin().await.with_context(|| "begin transaction")?;

    let mut _count = 0;
    for chunk in contests.chunks(1000) {
        _count += insert_contests(&mut tx, chunk)
            .await
            .with_context(|| "insert contests")?;
    }

    tx.commit().await.with_context(|| "commit transaction")?;

    Ok(())
}

async fn insert_contests<'a, A>(db: A, contests: &[Contest]) -> anyhow::Result<u64>
where
    A: Acquire<'a, Database = Postgres>,
{
    if contests.is_empty() {
        return Ok(0);
    }

    let mut builder: QueryBuilder<Postgres> = sqlx::QueryBuilder::new(
        r#"INSERT INTO "contests" ("contest_id", "start_epoch_second", "duration_second", "title", "rate_change", "category", "updated_at") "#,
    );
    builder.push_values(contests.iter(), |mut separated, c| {
        separated
            .push_bind(&c.id)
            .push_bind(&c.start_epoch_second)
            .push_bind(&c.duration_second)
            .push_bind(&c.title)
            .push_bind(&c.rate_change)
            .push_bind(c.categorize())
            .push("NOW()");
    });
    builder.push(r#" ON CONFLICT ("contest_id") DO UPDATE SET "contest_id" = EXCLUDED."contest_id", "start_epoch_second" = EXCLUDED."start_epoch_second", "duration_second" = EXCLUDED."duration_second", "title" = EXCLUDED."title", "rate_change" = EXCLUDED."rate_change", "category" = EXCLUDED."category", "updated_at" = NOW()"#);

    let mut conn = db.acquire().await.with_context(|| "acquire connection")?;

    let res = builder
        .build()
        .execute(&mut *conn)
        .await
        .with_context(|| "execute insert contests query")?;

    Ok(res.rows_affected())
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::testutil::{create_container, create_pool_from_container};
    use testcontainers::runners::AsyncRunner;

    #[tokio::test]
    async fn test_insert_contests() {
        let container = create_container().unwrap().start().await.unwrap();
        let pool = create_pool_from_container(&container).await.unwrap();

        // test insert empty
        let contests = vec![];
        let count = insert_contests(&pool, &contests).await.unwrap();
        assert_eq!(count, 0);

        // test insert single contest
        let contests = vec![Contest {
            id: String::from("abc001"),
            start_epoch_second: 1468670400,
            duration_second: 6000,
            title: String::from("abc001"),
            rate_change: String::from("-"),
        }];
        let count = insert_contests(&pool, &contests).await.unwrap();
        assert_eq!(count, 1);

        // test insert multiple contests
        let contests = vec![
            Contest {
                id: String::from("abc001"),
                start_epoch_second: 1468670400,
                duration_second: 6000,
                title: String::from("abc001"),
                rate_change: String::from("-"),
            },
            Contest {
                id: String::from("abc002"),
                start_epoch_second: 1468670400,
                duration_second: 6000,
                title: String::from("abc002"),
                rate_change: String::from("-"),
            },
        ];
        let count = insert_contests(&pool, &contests).await.unwrap();
        assert_eq!(count, 2);
    }
}
